### PR TITLE
[Core] Allow specifying custom Executor

### PR DESCRIPTION
**Pull request by Yard1**
This PR allows the user to pass in their custom Executor and TokenizerGroup classes if they so desire. This allows easy experimentation and customized deployments.

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Review by comaniac — APPROVED** [review]
LGTM!

File: vllm/executor/executor_base.py [hunk 1]
```diff
@@ -18,6 +18,8 @@ class ExecutorBase(ABC):
     that can execute the model on multiple devices.
     """
 
+    uses_ray: bool  # whether the executor uses Ray for orchestration.
```
Comment (rkooo567) [inline vllm/executor/executor_base.py:21]: Consider making it a property abstract class? That way, every subclass should implement this?

**Review by rkooo567 — APPROVED** [review]
I prefer to make uses_ray a property + abstract class, but other than that it lgtm

**Comment by njhill** [comment]
@Yard1 I'm kind of wary about making this pluggable since it's something additional that needs to be supported and may hamper our ability to evolve the architecture of these parts. Are there concrete examples / use cases of this that can be shared?

If it's just for the purpose of experimentation, it's not difficult to do that with custom changes in a separate branch?

**Comment by comaniac** [comment]
> @Yard1 I'm kind of wary about making this pluggable since it's something additional that needs to be supported and may hamper our ability to evolve the architecture of these parts.

I consider this is not a big issue because of the following reasons:
1. This PR only provides a more flexible way to accept custom executors without changing any existing behaviors.
2. As mentioned in the PR description, it's mainly for development convenience, meaning that we won't encourage users to implement their custom executors. It implies that custom executors shouldn't block future architecture evolution, as custom executors are not upstreamed, after all.

> Are there concrete examples / use cases of this that can be shared?
> If it's just for the purpose of experimentation, it's not difficult to do that with custom changes in a separate branch?

I could provide 2 use cases as examples:
1. We may need to develop a feature in executor/worker/model runner that would take a few weeks. However, since executor/worker/runner are the core components in vLLM and frequently changed/updated. As a result, we suffer from merging conflicts, and we have 2 options to solve this: (1) do not merge main branch until we finish developing; (2) use custom executor during the development process. We don't prefer (1) because it's hard to perform integration testing along with new feature in the other components of the upstream.
2. We may have a custom executor/worker/model runner that cannot be upstreamed easily, mainly due to a temporary hacky solution for particular business use cases. In this case, our upstream steps are usually (1) delivering business needs; (2) making the solution less hacky and more general; (3) propose an RFC to the community; (4) upstreaming the code. This process may take a few months, and it would be super tedious if we use a separate branch.

Thanks.

File: vllm/executor/executor_base.py (hunk 1 above)
Comment (Yard1) [reply vllm/executor/executor_base.py:21]: this should already function like that (it will raise exceptions if not set in implementation)

**Comment by njhill** [comment]
@comaniac @Yard1 ok fair enough! I guess we should clearly document that this isn't for general use though and could change/break at any point in future.

> It implies that custom executors shouldn't block future architecture evolution, as custom executors are not upstreamed, after all.

I meant in terms of us having to avoid changing the executor structure such that it would break others' custom executors.

> We may need to develop a feature in executor/worker/model runner that would take a few weeks. However, since executor/worker/runner are the core components in vLLM and frequently changed/updated. As a result, we suffer from merging conflicts

I thought that this was vLLM BAU :-) I'm not sure how the executor is special in this regard. Any large change that takes time to get merged ends up having to be regularly kept up-to-date with main (which I agree is very painful). We've had some PRs that took multiple months to get merged and required countless nontrivial conflict resolutions over that time (for example the multiproc executor, soft prompt support, ...)

**Comment by comaniac** [comment]
> @comaniac @Yard1 ok fair enough! I guess we should clearly document that this isn't for general use though and could change/break at any point in future.

👍 

> I meant in terms of us having to avoid changing the executor structure such that it would break others' custom executors.

Understood. IMO it's not the OSS community's responsibility to make sure the change of executor structure won't break other's downstream custom executors.

> I thought that this was vLLM BAU :-) I'm not sure how the executor is special in this regard. Any large change that takes time to get merged ends up having to be regularly kept up-to-date with main (which I agree is very painful). We've had some PRs that took multiple months to get merged and required countless nontrivial conflict resolutions over that time (for example the multiproc executor, soft prompt support, ...)

I totally understand, and that's also the motivation of this PR. Hopefully this could make our life less painful lol

**Comment by Yard1** [comment]
Added a comment to arg_utils!

**Pull request merged**